[007]: Add an nsinit demo launching a container from IPFS

wking opened the pull request:
Without bothering with Docker's image format, layer tarballs, etc.
Details in the demo's README.md.  I'm happy to rebase around #19
(which is currently also using the 005 number) if that PR lands first.

wking committed:
[007]: Add an nsinit demo launching a container from IPFS

Without bothering with Docker's image format, layer tarballs, etc.

wking committed:
[007]: Link to the root filesystem with http://gateway.ipfs.io/ipfs/

Make the IPFS key less mysterious ;).

jbenet commented:
@wking may be easier to turn this into `[006]`. its ok if the demos on the repo are weird numbers. we can reorder them later.

jbenet commented on 005-nsinit-container-from-ipfs-filesystem/README.md:
```diff
@@ -0,0 +1,98 @@
+file permissions][no-exec], so you can't launch `nsinit` directly from
+your IPNS directory.  As a workaround, the demo `Makefile` currently
+extracts the IPFS directory to your local filesystem, and then tweaks
+the file mode locally.
```
@whyrusleeping maybe we can add perms to unixfs soon? we should consider it for sure in next week's sprint. https://github.com/ipfs/pm/issues/5

jbenet commented on 005-nsinit-container-from-ipfs-filesystem/README.md:
```diff
@@ -0,0 +1,98 @@
+        $ go get -u github.com/ipfs/go-ipfs/cmd/ipfs
+
+   But `nsinit` is [a bit more complicated due to
+   vendoring][libcontainer-vendoring].  I had success with:
```
maybe we could make a simple bundle and add it to ipfs? i.e. say, this is how to do it for recent versions: `...` but if you want a frozen version that works, try: `ipfs get -o=libcontainer /ipfs/<hash>`

wking replied:
On Thu, Apr 16, 2015 at 06:02:58PM -0700, Juan Batiz-Benet wrote:

> maybe we could make a simple bundle and add it to ipfs?

That's a good idea, but:
1. I think we want a static binary, and I've had trouble creating
   static binaries with Go (but I don't have much Go experience either
   ;).
2. We can't just have one binary.  We probably want
   /ipns/…/libcontainer/v1.4/${GOOS}/${GOARCH}/nsinit.  (Note to self: you
   can list your system's default GOOS, GOARCH, … with ‘go env’).
3. It would be nice if we could talk the libcontainer folks into
   releasing official nsinit static binaries, since they understand
   their build system better than I do.

wking replied:
On Fri, Apr 17, 2015 at 08:51:33AM -0700, W. Trevor King wrote:

> Thu, Apr 16, 2015 at 06:02:58PM -0700, Juan Batiz-Benet:
> 
> > maybe we could make a simple bundle and add it to ipfs?
> 
> That's a good idea, but:
> 1. I think we want a static binary…

I gave up on the static binary and just bundled libc, libpthread, and
ld into my demo filesystem.  I added my local build of ‘ipfs’ and
‘nsinit’ for now, but it would be nice to be able to use official
upstream releases.  Since it's a Linux-container demo, you'll have to
be running on Linux anyway (possibly via a VM), so the
Linux/amd64-specific binaries are probably not too much of an issue
(although support for other arches would be nice).

Anyhow, all you need for the current iteration of the demo is a local
‘ipfs’ daemon.

jbenet commented:
@wking this looks great. i havent tried it yet, but maybe i can try it tomorrow.

wking commented:
On Thu, Apr 16, 2015 at 05:59:26PM -0700, Juan Batiz-Benet wrote:
> @wking may be easier to turn this into `[006]`. its ok if the demos
> on the repo are weird numbers. we can reorder them later.

Maybe I should make it [101], since it's the first demo in a
nsinit-based track, which is pretty independent of @hosh's existing
virtual machine and Docker-based demos.  Or maybe [0xx] for VMs, [1xx]
for Docker, and [2xx] for nsinit, where the hundreds digit is just the
order in which they arrived.  I'm happy to use a different hundreds
ordering if you'd prefer something else.

wking committed:
[007]: README.md: Launch ipfs daemon

Without this running you can still 'ipfs get ...' local objects, but
remote objects will give you:

  $ ipfs ls QmYRpzwXkSnCPPNL4b2gn5AFnyyMHFxhS9Nrbau5ycw92H
  Error: blockstore: block not found

wking committed:
[007]: Bundle ipfs and nsinit in the demo filesystem

You can still build it from source if you want, but this lets you skip
that step if you'd rather.  I've also included some of the basic
libraries from the Gentoo stage3 so we can handle a bit of dynamic
linking in Go binaries.

wking committed:
[007]: Add container.json to the IPFS directory

This is one less step folks need to run to launch from the directory,
although it does mean we have something else to update if the
libcontainer folks change their default config format, or tweak the
default values too far.

Conflicts:
	007-nsinit-container-from-ipfs-filesystem/README.md

wking commented:
After discussion yesterday, I've renumbered this to 007.

wking commented:
On Mon, Apr 20, 2015 at 01:13:03PM -0700, Juan Batiz-Benet wrote:
> Assigned #22 to @wking.

I'm not sure what I'm getting assigned to do here ;).  I'll work on
getting access mode metadata in place, and that will remove one of the
current pain points here.
